task_4/2.js

```diff
@@ -0,0 +1,111 @@
+// Реализовать 2 разных типа верфей, каждая из которых должна будет специализироваться на своем типе кораблей
+//
+// Для моторных кораблей доступны следующие специфичные характеристики:
+//   Мощность двигателя
+//   Материал корпуса
+//
+// Для парусных кораблей доступны следующие специфичные характеристики:
+//   Количество мачт
+//   Общая площадь парусов
+//
+// Что можно делать в верфи:
+//   Строить корабли - Должен проверяться тип корабля, работать только с кораблями своего типа
+//   Ремонтировать корабли - Должен проверяться тип корабля, работать только с кораблями своего типа
+//   Перекрашивать корабли - Можно красить любые корабли
+//   Обменивать старый корабль на новый - Можно обменивать только корабли того же типа
+//
+// Главное требование: Чистый код, весь дублирующийся код для кораблей и верфей должен быть вынесен в их прототипы.
+// Задание потребует воспользоваться гуглом для решения.
+//
+// Верфи и корабли должны создаваться с помощью функций-конструкторов.
+
+
+const shipyardPrototype = {
+  kind: {
+    configurable: false,
+    get: function() {
+      return this.shipyardKind
+    }
+  },
+  changeShipColor: function(ship, color) {
+    ship.color = color
+  },
+  createShip: function(kind, ...args) {
+    return this.handleShip(['kind', kind],function() {
+      switch (kind) {
+        case 'motor':
+          return new MotorShip(...args)
+        case 'sail':
+          return new SailShip(...args)
+      }
+    })
+  },
+  repairShip: function(ship) {
+    this.handleShip(['ship', ship], function() { console.log('repairing...') })
+  },
```

Comment:
И также можно немного упростить логику проверки типа корабля:
```suggestion
  repairShip: function(ship) {
    if (this.checkAccess(ship)) console.log('repairing...')
  },
```